app/namespaces/test_adjustment/test_adjustment_service.py

```diff
@@ -0,0 +1,267 @@
+# monte-carlo version of RG estimator,
+# accounting for uncertainty in se and sp estimates
+import pickle
+from math import log
+from hashlib import md5
+from typing import Dict, Tuple, Union
+
+import pandas as pd
+import pystan
+import arviz
+from statsmodels.stats.proportion import proportion_confint
+from marshmallow import Schema, fields, ValidationError
+
+from app.utils.test_adjustment import bastos_estimates, testadj_model_code
+
+
+def logit(p, tol=1e-3):
+    # logit function; well defined for p in open interval (0,1)
+
+    # constrain the probability to the range [tol, 1-tol]
+    # to avoid NaNs if p = 0 or p = 1
+    p = max(p, tol)
+    p = min(p, 1 - tol)
+
+    # return the logit of the constrained probability
+    return log(p / (1 - p))
+
+
+def result_is_bounded(median_adj_prev, raw_prev):
+    # check whether the adjusted result is close to the raw value
+    # e.g., log odds within 1
+    logit_delta = logit(median_adj_prev) - logit(raw_prev)
+    adjusted_closeto_raw = abs(logit_delta) < 1
+
+    # permit the adjusted result if adjusted and raw are both < 0.5
+    # or if both are > 0.5
+    both_below_maxsmall = (median_adj_prev <= 0.5) and (raw_prev <= 0.5)
+    both_above_minbig = (median_adj_prev >= 0.5) and (raw_prev >= 0.5)
+
+    return (adjusted_closeto_raw or both_below_maxsmall or both_above_minbig)
+
+
+def validate_against_schema(input_payload, schema):
+    try:
+        payload = schema.load(input_payload)
+    except ValidationError as err:
+        return err.messages
+    return payload
+
+
+class ModelParamsSchema(Schema):
+    n_prev_obs = fields.Integer()
+    y_prev_obs = fields.Float()
+    n_se = fields.Integer()
+    y_se = fields.Float()
+    n_sp = fields.Integer()
+    y_sp = fields.Float()
+
+
+class TestAdjHandler:
+    def __init__(self, model_code=testadj_model_code, model_name='testadj_binomial_se_sp',
+                 execution_params={}):
+        self.TESTADJ_MODEL = self.get_stan_model_cache(model_code=model_code, model_name=model_name)
+        self.TESTADJ_MODEL_NAME = model_name
+        # Parse execution params
+        self.n_iter = execution_params.get('n_iter', 4000)
+        self.n_chains = execution_params.get('n_chains', 3)
+        self.return_fit = execution_params.get('return_fit', False)
+        self.n_replicates = execution_params.get('n_replicates', 5)
```

Comment:
Okay, so. In the first instance, the algorithm that we were using was as follows:

- We were running the model 5x, and comparing the results between those 5 runs to see if it was consistent
- If it was consistent, we were taking the median result
- If it was not consistent, we were throwing it away and trying again - up to a total of 100 trials total, or 500 total model runs

This is a huge driver of why the test adjustment code was so slow. 

If we integrate some of the suggested optimizations, we should get a major, major speedup, that means that we can change our algorithm to the following:

- Run one model
- See whether it returns a sane result, with `result_is_bounded` and per the HMC diagnostics 
- If not, try running it again, up to a maximum of five models

I have highlighted the places where we can drop or simplify code in order to achieve this line by line below 
As a first step relevant to lines 69 and 70, we now no longer need the n_replicates param, and the default for trials_lim should be 5